pil-analyzer/src/traits_resolver.rs

```diff
@@ -0,0 +1,82 @@
+use powdr_ast::{
+    analyzed::{Expression, PolynomialReference},
+    parsed::{
+        types::{TupleType, Type},
+        TraitImplementation,
+    },
+};
+use std::{collections::HashMap, sync::Arc};
+
+use crate::type_unifier::Unifier;
+
+type SolvedImpl = ((String, Vec<Type>), Arc<Expression>);
+
+pub struct TraitsResolver<'a> {
+    trait_impls: &'a HashMap<String, Vec<TraitImplementation<Expression>>>,
+    solved_impls: HashMap<String, HashMap<Vec<Type>, Arc<Expression>>>,
+}
+
+impl<'a> TraitsResolver<'a> {
+    pub fn new(trait_impls: &'a HashMap<String, Vec<TraitImplementation<Expression>>>) -> Self {
+        Self {
+            trait_impls,
+            solved_impls: HashMap::new(),
+        }
+    }
+
+    pub fn resolve_trait(&mut self, ref_poly: &PolynomialReference) -> Result<(), String> {
+        if let Some(inner_map) = self.solved_impls.get(&ref_poly.name) {
+            if let Some(type_args) = &ref_poly.type_args {
```

Comment:
If ref_poly.type_args is None, we can return right away at the start of the function because it's not a trait function reference.